backend/groth16/bn254/solidity.go

```diff
@@ -490,18 +632,77 @@ contract Verifier {
     /// proof was successfully verified.
     /// @param proof the points (A, B, C) in EIP-197 format matching the output
     /// of compressProof.
+    {{- if gt $numCommitments 0 }}
+    /// @param commitments the Pedersen commitments from the proof.
+    /// @param commitmentPok the proof of knowledge for the Pedersen commitments.
+    {{- end }}
     /// @param input the public input field elements in the scalar field Fr.
     /// Elements must be reduced.
     function verifyProof(
         uint256[8] calldata proof,
-        uint256[{{$numPublic}}] calldata input
+        {{- if gt $numCommitments 0}}
+        uint256[{{mul 2 $numCommitments}}] calldata commitments,
+        uint256[2] calldata commitmentPok,
+        {{- end }}
+        uint256[{{$numWitness}}] calldata input
     ) public view {
+        {{- if eq $numCommitments 0 }}
         (uint256 x, uint256 y) = publicInputMSM(input);
+        {{- else }}
+        // HashToField
+        uint256[{{$numCommitments}}] memory publicCommitments;
+        {{- range $i := intRange $numCommitments }}
+            publicCommitments[{{$i}}] = uint256(
+                sha256(
+                    abi.encodePacked(
+                        commitments[{{mul $i 2}}],
```

Comment:
On larger witness data, this will result in `Stack too deep` or `memoryguard was present` with `via-ir`. 
Should try and pack the data using assembly to prevent this. Will see if I can contribute the necessary code.

Reply:
I'm not too knowledgeable about Solidity to comment on that. But I think in general it doesn't make sense to have too much of public witness as it increases MSM size, making the verifier more expensive.

Rather, what the circuit designer could do is that for expected public witness `P1, ..., Pn` it provides as a public witness only `P* = Hash(P1, ..., Pn)` and then as a private witness `P1, ..., Pn`. Then, in circuit it asserts that `P* = Hash(P1, ..., Pn)` holds. Now, when we want to actually use the values `P1, .., Pn` in Solidity, then we would have to perform same hashing in Solidity to verify that public witness `P*` and values `P1, ..., Pn` correspond to each other.

NB! Keep in mind that we would have to use ZK friendly hash function for that in order for the in-circuit assertion to be efficient.

Reply:
One approach would be to try to use some other hash function or try some continues hashing with several calls, but dunno if Solidity allows this.

But I think that in this PR we can keep as is.